[1.18.2] Add tree decay back to Palms, Mahognay and Home Tree

Dragon-Seeker opened the pull request:
Basically adds leaf decay for the big trees by diagonally checking for leaf blocks

Dragon-Seeker committed:
Init Commit for code

Gegy commented on src/main/java/net/tropicraft/core/common/block/experimental/TropicraftExperimentalLeaveBlock.java:
```diff
@@ -0,0 +1,117 @@
+        int currentExtraDistance = 20; //*
+        int length = directionDiagonals.length;
+
+        for(int i = 0; i < length; ++i) {
```
Random thought: maybe it's worth doing a `for (BlockPos pos : BlockPos.betweenClosed(pos.offset(-1, -1, -1), pos.offset(1, 1, 1))` kind-of-thing instead of having the `DiagonalDirections`? Can use `BlockPos.distManhattan()` to find the `additionAmount` 🙂

Dragon-Seeker replied:
At the time I had no experience with BlockPos and should have asked around for some better ways to do this but yea, that would work a lot better than the hard coded directions. Also did not know that Manhatten Distance was a thing at all nor what its meaning was.

Gegy commented on src/main/java/net/tropicraft/core/common/block/experimental/TropicraftExperimentalLeaveBlock.java:
```diff
@@ -0,0 +1,117 @@
+    @Override
+    public boolean isRandomlyTicking(BlockState state) { //*
+        //Checks if it is set Persistent but also checks for the custom leaf distance
+        return state.getValue(EXTRADISTANCE) == 20 && !(Boolean)state.getValue(PERSISTENT);
```
Nit: would it be worth extracting this `20` into a constant/function? 🙂

Dragon-Seeker replied:
Sounds like a good idea

Gegy commented on src/main/java/net/tropicraft/core/common/block/experimental/TropicraftExperimentalLeaveBlock.java:
```diff
@@ -0,0 +1,117 @@
+
+    public TropicraftExperimentalLeaveBlock(Properties settings) {
+        super(settings);
+        this.registerDefaultState(this.defaultBlockState().setValue(DISTANCE, 7).setValue(EXTRADISTANCE, 20).setValue(TEMPIGNORERANDOMTICK, false));
```
Would it be possible to entirely replace the existing `DISTANCE` property instead of having both? Mainly imagining this would otherwise end up with a full 20\*7\*2 block states, which is quite a lot 😄

Dragon-Seeker replied:
At the time, I was worrired that due to the class inheriting the Base Leaf Block, It could cause issue if somewhere internally the block was then set with DISTANCE property causing a crash or exception due to it missing such

Dragon-Seeker replied:
Will need to make sure that this doesn't cause any Major Incompatibility with other mods attempting to get DISTANCE data. Maybe I could somehow interpreted the call to get the DISTANCE from a TropicratExperimentalLeaveBlock (Better name is needed), as wanting the CUSTOM_DISTANCE value just interpolated to 1-7 scale instead of 1-20?

Gegy commented on src/main/java/net/tropicraft/core/common/block/experimental/TropicraftExperimentalLeaveBlock.java:
```diff
@@ -0,0 +1,117 @@
+
+public class TropicraftExperimentalLeaveBlock extends LeavesBlock {
+
+    public static IntegerProperty EXTRADISTANCE;
```
Do the palm trees need an extra distance here, or would it be possible to entirely re-use the 0-7 range here? If possible, it might help cut down on blockstates to have different ranges depending on the tree size 🙂

Dragon-Seeker replied:
Yes, some of the more widely shaped ones are pretty big due to the amount of diagonale blocks and the huge palm tree for sure needs it. Ill double check this one later to just confirm what I remember.

Dragon-Seeker replied:
We could set the MAX_DISTANCE value for each block to shave off even more blockstates since:

- Palm trees max distance it could be away from a log would be 11 (12 as the decay distance)
- Kapok (Rainforest trees / Mahogany) distance would be more like 20 (21 as the decay distance) 
- Mahogany(From the home tree which still is weird) distance would be more like 18 (19 as the decay distance)

I don't think papaya trees need this based on the trees leaves shape

Dragon-Seeker replied:
Well I had to do some cursed f***ing things for this to work ngl but here is the cursed code:

``` java
public class TropicraftExperimentalLeaveBlock extends LeavesBlock {

    private final int maxDistanceBeforeDecay;

    public IntegerProperty CUSTOM_DISTANCE;

    //Property to check if there has even been a block update to prevent unwanted decay instantly
    public static BooleanProperty TEMPORARY_IGNORE_DECAY = BooleanProperty.create("temp_ignore_decay");

    public TropicraftExperimentalLeaveBlock(Properties settings, int maxDistance){
        super(settings);

        //Cursedness inbound
        this.CUSTOM_DISTANCE = IntegerProperty.create("custom_distance", 1, maxDistance);
        this.maxDistanceBeforeDecay = maxDistance;

        StateDefinition.Builder<Block, BlockState> builder = new StateDefinition.Builder<>(this);

        this.createBlockStateDefinition(builder);
        builder.add(CUSTOM_DISTANCE);

        ObfuscationReflectionHelper.setPrivateValue(Block.class, this, builder.create(Block::defaultBlockState, BlockState::new), "f_49792_");

        this.registerDefaultState(this.stateDefinition.any()
                .setValue(PERSISTENT, false)
                .setValue(CUSTOM_DISTANCE, maxDistanceBeforeDecay)
                .setValue(TEMPORARY_IGNORE_DECAY, false));
    }
...
}
```

Gegy reviewed (commented):
I think you mentioned this before, but totally forgot, so asking again 😅 - would it be possible to set the distance values at generation time, too? I imagine there would be quite a lot of block updates that come from these, especially with e.g. the home tree 😄 

In the case of the home tree though, the structure files probably need an update to have the correct blockstate values 🙂

Dragon-Seeker commented:
> I think you mentioned this before, but totally forgot, so asking again 😅 - would it be possible to set the distance values at generation time, too? I imagine there would be quite a lot of block updates that come from these, especially with e.g. the home tree 😄

I had issues with it not being set properly which could come down to how the leaves are placed and the values not updating properly causing the values to all be 20 for some time until the blocks update causing some Swiss cheesing of the leaves parasitically noticeable on the big mahogany trees.

I'm wondering if this is something to do with the Tree Feature ?

Dragon-Seeker committed:
Remove need for Diagonal Directions Enum and overall clean up and reformat class

Dragon-Seeker committed:
Fix Certain Crashes internally within the LeavesBlock

Dragon-Seeker committed:
Cursed way of cutting down on blockstates got to say

This may be undone if found to be too cursed

Dragon-Seeker committed:
Attempt Better Level of cursed

This removes the need to set the stateDefinition again but uses a static int to store the maxDistanceBeforeDecay value until createBlockStateDefiniton is called within Block Class Constructor

Dragon-Seeker committed:
Fix need for Custom Boolean Property due to Decay issues

Was caused by distance values not being set properly. Basically a copy of vanillas methods but using the CUSTOM_DISTANCE property and maxDistance value from the given Experimental Tropicraft Leaf Block

Dragon-Seeker commented:
Latest Commit should be alot less cursed and with proper Block Updates for leaves

Dragon-Seeker commented:
Though the method for the Palm Trees and Mahogany Trees didn't work on the Home trees leave branches so that Still needs to be fixed

NotActuallyTerry commented:
@Gegy where are we with this?
